draft-irtf-cfrg-vdaf.md

```diff
@@ -2384,140 +2431,122 @@ def joint_rand(Prio3, k_joint_rand_seed):
-    if Prio3.Flp.JOINT_RAND_LEN == 0:
-        if len(encoded) != 0:
-            raise ERR_DECODE
-        return (meas_share, proof_share, None)
-    k_blind, encoded = encoded[:l], encoded[l:]
-    if len(encoded) != 0:
-        raise ERR_DECODE
-    return (meas_share, proof_share, k_blind)
+~~~
+struct {
+    Prio3Seed k_meas_share;
+    Prio3Seed k_proof_share;
+} Prio3HelperShare;
 ~~~
 
-Next, the methods below are used for encoding and decoding the VDAF public share.
+When joint randomness is used, the Leader's input share is structured as
+follows:
 
 ~~~
-def encode_public_share(Prio3,
-                        k_joint_rand_parts):
-    encoded = Bytes()
-    if Prio3.Flp.JOINT_RAND_LEN > 0:
-        encoded += concat(k_joint_rand_parts)
-    return encoded
+struct {
+    Prio3LeaderShare inner;
+    Prio3Seed k_blind;
+} Prio3LeaderShareWithJointRand;
+~~~
 
-def decode_public_share(Prio3, encoded):
-    l = Prio3.Prg.SEED_SIZE
-    if Prio3.Flp.JOINT_RAND_LEN == 0:
-        if len(encoded) != 0:
-            raise ERR_DECODE
-        return None
-    k_joint_rand_parts = []
-    for i in range(Prio3.SHARES):
-        k_joint_rand_part, encoded = encoded[:l], encoded[l:]
-        k_joint_rand_parts.append(k_joint_rand_part)
-    if len(encoded) != 0:
-        raise ERR_DECODE
-    return k_joint_rand_parts
+Finally, when joint randomness is used, the Helpers' shares are structured as
+follows:
+
+~~~
+struct {
+    Prio3HelperShare inner;
+    Prio3Seed k_blind;
+} Prio3HelperShareWithJointRand;
 ~~~
 
-Finally, the methods below are used for encoding and decoding the values
-transmitted during VDAF preparation.
+#### Prep Share
+
+When joint randomness is not used, the prep share is structured as follows:
 
 ~~~
-def encode_prep_share(Prio3, verifier, k_joint_rand):
-    encoded = Bytes()
-    encoded += Prio3.Flp.Field.encode_vec(verifier)
-    if Prio3.Flp.JOINT_RAND_LEN > 0:
-        encoded += k_joint_rand
-    return encoded
+struct {
+    Prio3.Flp.Field verifier_share[Prio3.Flp.VERIFIER_LEN];
+} Prio3PrepShare;
+~~~
 
-def decode_prep_share(Prio3, encoded):
-    l = Prio3.Flp.Field.ENCODED_SIZE * Prio3.Flp.VERIFIER_LEN
-    encoded_verifier, encoded = encoded[:l], encoded[l:]
-    verifier = Prio3.Flp.Field.decode_vec(encoded_verifier)
-    if Prio3.Flp.JOINT_RAND_LEN == 0:
-        if len(encoded) != 0:
-            raise ERR_DECODE
-        return (verifier, None)
-    l = Prio3.Prg.SEED_SIZE
-    k_joint_rand, encoded = encoded[:l], encoded[l:]
-    if len(encoded) != 0:
-        raise ERR_DECODE
-    return (verifier, k_joint_rand)
+When joint randomness is used, the prep share includes the Aggregator's joint
+randomness part and is structured as follows:
 
-def encode_prep_msg(Prio3, k_joint_rand_check):
-    encoded = Bytes()
-    if Prio3.Flp.JOINT_RAND_LEN > 0:
-        encoded += k_joint_rand_check
-    return encoded
+~~~
+struct {
+    Prio3.Flp.Field verifier_share[Prio3.Flp.VERIFIER_LEN];
```

Comment:
nit: here is another opportunity to reuse `Prio3PrepShare`, similar to `Prio3LeaderShare`.